vertexy/src/private/constraints/ShortestPathConstraint.cpp

```diff
@@ -0,0 +1,596 @@
+{
+	m_lastValidTimestamp.clear();
+}
+
+bool ShortestPathConstraint::isValidDistance(const IVariableDatabase* db, int dist) const
+{
+	switch (m_op)
+	{
+	case EConstraintOperator::GreaterThan:
+		return dist > db->getMinimumPossibleDomainValue(m_distance);
+	case EConstraintOperator::GreaterThanEq:
+		return dist >= db->getMinimumPossibleDomainValue(m_distance);
+	case EConstraintOperator::LessThan:
+		return dist < db->getMaximumPossibleDomainValue(m_distance);
+	case EConstraintOperator::LessThanEq:
+		return dist <= db->getMaximumPossibleDomainValue(m_distance);
+	}
+
+	vxy_assert(false); //NotEqual not supported
+	return false;
+}
+
+//is possibly reachable
+bool ShortestPathConstraint::isPossiblyValid(const IVariableDatabase* db, const ReachabilitySourceData& src, int vertex) 
+{
+	// we do not need to check for reachability here because we always call isPossiblyReachable before calling this
+	if (m_op == EConstraintOperator::GreaterThan || m_op == EConstraintOperator::GreaterThanEq)
+	{
+		if (src.minReachability->isReachable(vertex))
+		{
+			if (!isValidDistance(db, src.minReachability->getDistance(vertex)))
+			{
+				// the shortest path is too short, so our vertex is definitely unreachable (ie it cannot be a target)
+				removeTimestampToCommit(db, src.vertex, vertex);
+				return false;
+			}
+			else
+			{
+				addTimestampToCommit(db, src.vertex, vertex);
+				return true;
+			}
+		}
+		else if (src.maxReachability->isReachable(vertex))
+		{
+			return true;
+		}
+	}
+	else
+	{
+		if (src.minReachability->isReachable(vertex) && isValidDistance(db, src.minReachability->getDistance(vertex)))
+		{
+			return true;
+		}
+		else if (src.maxReachability->isReachable(vertex))
+		{
+			if (!isValidDistance(db, src.maxReachability->getDistance(vertex)))
+			{
+				removeTimestampToCommit(db, src.vertex, vertex);
+				return false;
+			}
+
+			addTimestampToCommit(db, src.vertex, vertex);
+			return true;
+		}
+	}
+	return false;
+}
+
+ITopologySearchConstraint::EValidityDetermination ShortestPathConstraint::determineValidityHelper(
+	const IVariableDatabase* db,
+	const ReachabilitySourceData& src,
+	int vertex,
+	VarID srcVertex) 
+{
+	if (m_op == EConstraintOperator::GreaterThan || m_op == EConstraintOperator::GreaterThanEq)
+	{
+		if (src.minReachability->isReachable(vertex))
+		{
+			if (!isValidDistance(db, src.minReachability->getDistance(vertex)))
+			{
+				// the shortest path is too short, so our vertex is definitely unreachable (ie it cannot be a target)
+				removeTimestampToCommit(db, src.vertex, vertex);
+				return EValidityDetermination::DefinitelyInvalid;
+			}
+
+			addTimestampToCommit(db, src.vertex, vertex);
+			if (definitelyIsSource(db, srcVertex))
+			{
+				return EValidityDetermination::DefinitelyValid;
+			}
+			else
+			{
+				return EValidityDetermination::PossiblyValid;
+			}
+		}
+		else if (src.maxReachability->isReachable(vertex))
+		{
+			return EValidityDetermination::PossiblyValid;
```

Comment:
I was going to say that we could return DefinitelyValid here if the maxReachability distance is > requiredDistance and the source is definite, since maxReachability distance will only increase (since edges only get removed from it). 
But I guess that's not correct, because edges might get removed that make this totally unreachable.

This makes me consider whether we should treat unreachable in this constraint as "infinitely far away". My reasoning is that 1) that might sometimes be desired behavior, 2) it will be able to more tightly constrain variables that way, and 3) you can always add an additional ReachabilityConstraint to ensure that it's both reachable AND > distance.

Reply:
That makes sense. It would only help the greater than case, and I can't think of a case where unreachable is what we want, though